Dockerfile

```diff
@@ -17,24 +18,16 @@ RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o lustre-csi-driver main.
 
 ENTRYPOINT ["/bin/sh"]
 
-#FROM builder as testing
-#WORKDIR /workspace
-#
-#COPY Makefile .
-#
-#RUN go install sigs.k8s.io/controller-runtime/tools/setup-envtest@latest && \
-#    make manifests && make generate && make fmt &&  make vet && \
-#    mkdir -p /workspace/testbin && /bin/bash -c "test -f /workspace/testbin/setup-envtest.sh || curl -sSLo /workspace/testbin/setup-envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/controller-runtime/v0.7.2/hack/setup-envtest.sh" && \
-#    /bin/bash -c "source /workspace/testbin/setup-envtest.sh; fetch_envtest_tools /workspace/testbin; setup_envtest_env /workspace/testbin"
-#
-#ENTRYPOINT ["bash", "/workspace/initiateContainerTest.sh"]
-
-# The final application stage.
-FROM redhat/ubi8-minimal
+# The final application stage, minimal image with compiled binary copied in
+# Note: The base image used for the container here is sensitive to the host OS
+# when performing lustre operations. For example, if running this container with
+# a RHEL base image, the CSI driver works well with a RHEL-compatible host OS
+# (i.e. CentOS), but may have issues with an OpenSUSE host OS. To resolve this,
+# bind-mounting host OS-compatible binaries into the container may be required.
+FROM registry.suse.com/bci/bci-minimal:latest
```

Comment:
Don't merge this until we've tested with Rabbit.

Reply:
It seems like it would be more consistent with the container paradigm to package a version of the Lustre utilities with the container.  That would avoid any userspace OS dependencies.  As for the user/kernel interface I'd expect those interface to be stable.

Reply:
I agree, but a drawback to this is that the container size then becomes much larger.

Reply:
I'm not sure that it needs to be "much" larger. If it is, then maybe we need to look at how we are building lustre. I think for the container we need to do a user-space tools build and skip all of the kernel module building.  We don't need to package kernel modules in the container image, and we don't want that to accidentally pull a full kernel into the container image, etc.

Reply:
For reference, a quick swag suggests we're talking about adding ~11Mb to the container size.  That doesn't seem too bad.

This is based on the RPM package size for Lustre and its dependencies, although I probably missed a few small ones.  I didn't include the kmod, debuginfo, and testing packages which we wouldn't need, and assumed that systemd, util-linux, glibc and openssl are all already there.  If not that adds about 20Mb.  Here's the rough breakdown.


size | rpm package
--- | ---
4032536 | libzpool5 
2014451 | lustre 
1674283 | zfs 
1473899 | sysstat 
589080 | libzfs5 
343952 | libblkid 
195735 | zlib 
169600 | libselinux 
135798 | libyaml 
95928 | libnvpair3 
80837 | libgudev 
69088 | libuutil3 
35088 | libuuid 
34252 | lustre-osd-zfs-mount 
__10944527__ | __total__

Reply:
Will be working on testing this today

Reply:
Those are the lustre *server* dependencies, which we don't want here; this container only needs lustre client functionality. Even though the lustre client doesn't need a lustre-patched kernel, it does still need to have user-space RPMs installed that are compatible with the host kernel. Here's the compatibility matrix: https://git.whamcloud.com/?p=fs/lustre-release.git;a=blob;f=lustre/ChangeLog

Per the [client installation instructions](https://wiki.lustre.org/Installing_the_Lustre_Software), the client packages installed should correspond to the kernel version installed. For example, installing a [lustre-2.12 el7](https://downloads.whamcloud.com/public/lustre/latest-2.12-release/el7/client/RPMS/) lustre-client RPM may not be compatible with a 5.3.18-lp152.81-default kernel.

This was one motivation for mounting/using the libraries/packages already installed on the host OS. Thoughts?

Reply:
Good point.  We could definitely prune this down much further if all we need is the Lustre client functionality.  Most of the client functionality is already being provided by the kmods on the host anyway.

For mounting all we need is the `mount.lustre` binary and any library dependencies (or a static build).  The mount syscall has been a stable interface forever so issues there seem very unlikely.  It's also my understanding that Lustre's policy is to avoid any breaking changes to the `lctl` ioctls.  It's of course possible something might sneak in, but that would be a bug to fix.  The /proc and /sys files do change more but as long as they're not being accessed directly that should be fine.  Are there any other Lustre interfaces needed, maybe the `llapi` library interface?

It seems like we could build a very lightweight version of the Lustre user space tools.

Reply:
Strictly speaking we don't need the "Lustre client" in the container image either, right? "Lustre client" refers to the linux kernel modules that implement a lustre filesystem on the client side. Like @behlendorf said we just need the user space tools in the container. The standard lustre spec file has an option to do a build without kernel modules, and then we wind up with just the user space parts in the rpm.

```rpmbuild --without lustre_modules lustre.spec```

Reply:
In my latest commit I've retrieved the `mount.lustre` user-space tool built for our Slingshot systems, and put this in the container. Testing without any of the `/usr/sbin, /usr/bin, /usr/lib, /usr/lib64, ...` etc volumeMounts works great, but that was just for a SLES host OS (in theory, it shouldn't matter) and an `@kfi` NID -- Dean's going to test this container on a Rabbit system with `@tcp` NID and RHEL host OS.

Reply:
Newest Dockerfile/image build works on Rabbit system as well, so this is not host OS dependent. I've pushed some more changes to the Makefile, Dockerfile, and config YAMLs, (see edited PR description) feel free to review them when you get a chance! Thanks